File: validator/sawtooth_validator/server/state.py
# Copyright 2016 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------
import logging
from sawtooth_validator.protobuf import state_context_pb2
from sawtooth_validator.server.message import Message


LOGGER = logging.getLogger(__name__)
LOGGER.addHandler(logging.StreamHandler())
LOGGER.setLevel(logging.DEBUG)


class GetHandler(object):

    def __init__(self, context_manager):
        self._context_manager = context_manager

    def handle(self, message, responder):
        get_request = state_context_pb2.GetRequest()
        get_request.ParseFromString(message.content)
        return_values = self._context_manager.get(get_request.context_id,
                                                  get_request.addresses)
        return_list = return_values if return_values is not None else []
        LOGGER.info("GET: %s", return_list)
        entry_list = [state_context_pb2.Entry(address=a,
                                              data=d) for a, d in return_list]
        responder.send(message=Message(
            correlation_id=message.correlation_id,
            message_type='state/getresponse',
            content=state_context_pb2.GetResponse(
                entries=entry_list).SerializeToString()
        ))


class SetHandler(object):
    def __init__(self, context_manager):
        """

        Args:
            context_manager (sawtooth_validator.context_manager.
            ContextManager):
        """
        self._context_manager = context_manager

    def handle(self, message, responder):
        LOGGER.info("We've got a set message")
        set_request = state_context_pb2.SetRequest()
        set_request.ParseFromString(message.content)
        set_values_list = [{e.address: e.data} for e in set_request.entries]
        return_value = self._context_manager.set(set_request.context_id,
                                                 set_values_list)
        response = state_context_pb2.SetResponse()
        if return_value is True:
            address_list = [e.address for e in set_request.entries]
            LOGGER.info("SET: %s", address_list)
            response.addresses.extend(address_list)
            responder.send(message=Message(
                correlation_id=message.correlation_id,
                message_type='state/setresponse',
                content=response.SerializeToString()
            ))
        else:
            response.addresses.extend([])
            responder.send(message=Message(
                correlation_id=message.correlation_id,
                message_type='state/setresponse',
                content=response.SerializeToString()
            ))

File: sdk/java/src/main/java/sawtooth/sdk/client/State.java
/* Copyright 2016 Intel Corporation
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
------------------------------------------------------------------------------*/

package sawtooth.sdk.client;

import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;

import sawtooth.sdk.processor.exceptions.InternalError;
import sawtooth.sdk.protobuf.Entry;
import sawtooth.sdk.protobuf.Message;
import sawtooth.sdk.protobuf.TpStateGetRequest;
import sawtooth.sdk.protobuf.TpStateGetResponse;
import sawtooth.sdk.protobuf.TpStateSetRequest;
import sawtooth.sdk.protobuf.TpStateSetResponse;


import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeoutException;


/**
 * Client state that interacts with the context manager through Stream networking.
 */
public class State {

  private Stream stream;
  private String contextId;

  public State(Stream stream, String contextId) {
    this.stream = stream;
    this.contextId = contextId;
  }

  /**
   * Make a Get request on a specific context specified by contextId.
   *
   * @param addresses a collection of address Strings
   * @return Map where the keys are addresses, values Bytestring
   * @throws InternalError something went wrong processing transaction
   */
  public Map<String, ByteString> get(Collection<String> addresses) throws InternalError {

    TpStateGetRequest getRequest = TpStateGetRequest.newBuilder()
            .addAllAddresses(addresses)
            .setContextId(this.contextId).build();
    FutureByteString future = stream.send(Message.MessageType.TP_STATE_GET_REQUEST,
        getRequest.toByteString());
    TpStateGetResponse getResponse = null;
    try {
      getResponse = TpStateGetResponse.parseFrom(future.getResult());
    } catch (TimeoutException toe) {
      try {
        getResponse = TpStateGetResponse.parseFrom(future.getResult());
      } catch (Exception e) {
        throw new InternalError(e.toString());
      }
    } catch (InterruptedException iee) {
      throw new InternalError(iee.toString());
    } catch (InvalidProtocolBufferException ipbe) {
      // server didn't respond with a GetResponse
      throw new InternalError(ipbe.toString());
    }
    Map<String, ByteString> results = new HashMap<String, ByteString>();
    if (getResponse != null) {
      for (Entry entry : getResponse.getEntriesList()) {
        results.put(entry.getAddress(), entry.getData());
      }
    }
    return results;
  }

  /**
   * Make a Set request on a specific context specified by contextId.
   *
   * @param addressValuePairs A collection of Map.Entry's
   * @return addressesThatWereSet, A collection of address Strings that were set
   * @throws InternalError something went wrong processing transaction
   */
  public Collection<String> set(Collection<java.util.Map.Entry<String,
          ByteString>> addressValuePairs) throws InternalError {
    ArrayList<Entry> entryArrayList = new ArrayList<Entry>();
    for (Map.Entry<String, ByteString> entry : addressValuePairs) {
      Entry ourEntry = Entry.newBuilder()
              .setAddress(entry.getKey())
              .setData(entry.getValue())
              .build();
      entryArrayList.add(ourEntry);
    }
    TpStateSetRequest setRequest = TpStateSetRequest.newBuilder()
            .addAllEntries(entryArrayList)
            .setContextId(this.contextId).build();
    FutureByteString future = stream.send(Message.MessageType.TP_STATE_SET_REQUEST,
        setRequest.toByteString());
    TpStateSetResponse setResponse = null;
    try {
      setResponse = TpStateSetResponse.parseFrom(future.getResult());
    } catch (TimeoutException toe) {
      try {
        setResponse = TpStateSetResponse.parseFrom(future.getResult());
      } catch (Exception e) {
        throw new InternalError(e.toString());
      }
    } catch (InterruptedException iee) {
      throw new InternalError(iee.toString());

    } catch (InvalidProtocolBufferException ipbe) {
      // server didn't respond with a SetResponse
      throw new InternalError(ipbe.toString());
    }
    ArrayList<String> addressesThatWereSet = new ArrayList<String>();
    if (setResponse != null) {
      for (String address : setResponse.getAddressesList()) {
        addressesThatWereSet.add(address);
      }
    }
    return addressesThatWereSet;
  }

}
